test/models/solution_test.rb

```diff
@@ -49,4 +41,48 @@ class SolutionTest < ActiveSupport::TestCase
     assert_equal solution.track.git_head_sha, solution.git_sha
     assert_equal solution.exercise.slug, solution.git_slug
   end
+
+  test "downloaded?" do
+    refute create(:concept_solution, downloaded_at: nil).downloaded?
+    assert create(:concept_solution, downloaded_at: Time.current).downloaded?
+  end
+
+  test "#completed?" do
+    refute create(:concept_solution, completed_at: nil).completed?
+    assert create(:concept_solution, completed_at: Time.current).completed?
+  end
+
+  test "#published?" do
+    refute create(:concept_solution, published_at: nil).published?
+    assert create(:concept_solution, published_at: Time.current).published?
+  end
+
+  test ".completed and .uncompleted" do
+    completed = create :concept_solution, completed_at: Time.current
+    not_completed = create :concept_solution, completed_at: nil
+
+    assert_equal [completed], Solution.completed
+    assert_equal [not_completed], Solution.not_completed
+  end
+
+  test ".for" do
+    user = create :user
+    exercise = create :concept_exercise
+    create :concept_solution, user: user
+    create :concept_solution, exercise: exercise
+    solution = create :concept_solution, user: user, exercise: exercise
+
+    assert_equal solution, Solution.for(user, exercise)
+  end
+
+  test "anonymised_user_handle" do
+    solution_1 = create :concept_solution
+    solution_2 = create :concept_solution
+
+    assert_equal solution_1.anonymised_user_handle,
+      "anonymous-#{Digest::SHA1.hexdigest("#{solution_1.id}-#{solution_1.uuid}")}"
```

Comment:
This is duplicating the actual implementation, but with the `id` and `uuid` not being stable, I don't think we can do much about that.

Reply:
Yeah - I could stub them both I guess. But I figured it was at least safe.